Feat code rag bench

boerz-coding opened the pull request:
# Description
I have moved my PR #2199 from my forked repo to the main repo, to allow proper chaining with follow-up PRs and CI testing.


This PR implements **Version 0.1 (MVP)** of the CodeRAG-Bench benchmark (issue #1462).

## Introduction
**CodeRAG-Bench** is designed to enable rigorous evaluations and advance research on retrieval-augmented code generation.

Reference: [CodeRAG-Bench Homepage](https://code-rag-bench.github.io/)

## Current Status (v0.1)

This MVP supports:
- **Canonical retrieval** and **generation** evaluation for the **HumanEval** sub-task.
- Saving retrieval results, generations, and evaluation metrics.
- Full retrieval-generation-evaluation pipeline for initial testing and feedback.

**Note:**  
The full benchmark includes 7 code generation tasks (HumanEval, MBPP, LiveCodeBench, DS-1000, ODEX, RepoEval, SWE-bench-Lite), and both canonical and open-corpus retrieval. This MVP focuses on the first task (HumanEval) with canonical retrieval only.

The goal of this MVP is to provide an early working version for feedback and for potential users to start experimenting.


## Example Output Files

Benchmark output files (retrieval results, generations, evaluation metrics) are available [here](https://github.com/boerz-coding/Camel_Coderag_bench_example_output/tree/main/V0.1MVP_example_output_full_retrieve_generate) for reproducibility.


## Next Steps (Mini Roadmap)

I will continue working towards the full CodeRAG-Bench implementation. I also plan to split tasks into sub-issues to invite community contributions whenever possible.

Planned next steps:
- [ ] Support open-corpus retrieval for HumanEval.
- [ ] Add support for the other 6 tasks (MBPP, LiveCodeBench, DS-1000, ODEX, RepoEval, SWE-bench-Lite) across retrieval and generation stages.
- [ ] (Optional) Add parallelization support.
- [ ] Write a Cookbook for quickstart and usage examples.




## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.

- [x ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have linked this PR to an issue using the Development section on the right sidebar or by adding `Fixes #issue-number` in the PR description (**required**)
- [x] I have checked if any dependencies need to be added or updated in `pyproject.toml` and `uv lock`
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation if needed:
- [x] I have added examples if this is a new feature

If you are unsure about any of these, don't hesitate to ask. We are here to help!

zjrwtx commented on camel/benchmarks/coderag_bench/code_generation_evaluation/code_eval.py:
thanks @boerz-coding ，as we talk in meeting,it seems that the origin code is apache protocol, so i guess we need to delete some long var etc

boerz-coding replied:
Done! Thank you!

boerz-coding commented:
Hi @zjrwtx I have edited all docstrings. Thank you for your helpful comments.

boerz-coding commented:
Hi @Wendong-Fan, I’ve addressed all comments from you and Yifeng — the PR is ready for review again.

Since this PR was moved from  my forked repo to the main repo and the your comments was left on the original PR , I wanted to recap the key changes from previous reviews:

- **Polished docstrings** across the core benchmark class and methods.
- **Applied lazy loading** for optional packages like `beir`, to reduce unnecessary imports.
- **Refactored the content of `CodeRAGBenchmark.run()`** into multiple sub-functions for better readability and modularity.
- **Updated retriever interface**: `CodeRAGBenchmark.run()` now accepts any callable retriever instead of requiring a specific retriever type.
